docs/topics/multiplatform/multiplatform-hierarchy.md

```diff
@@ -0,0 +1,108 @@
+
+```properties
+kotlin.mpp.enableGranularSourceSetsMetadata=true
+kotlin.native.enableDependencyPropagation=false
+```
+
+## For multiplatform project authors
+
+With the new hierarchical project structure support, you can share code among some, but not
+all [targets](multiplatform-dsl-reference.md#targets) in a multiplatform project.
+
+You can also use platform-dependent libraries, such as `UIKit`, and `posix` in source sets shared among several native
+targets. One popular case is having access to iOS-specific dependencies like `Foundation` when sharing code across all
+iOS targets. New structure helps you share more native code without being limited by platform-specific dependencies.
+
+By using the hierarchical structure along with platform-dependent libraries in shared source sets, you can eliminate the
+need to use certain workarounds to get IDE support for sharing source sets among several native targets, for
+example `iosArm64` and `iosX64`:
+
+```kotlin
+kotlin {
+    // workaround 1: select iOS target platform depending on the Xcode environment variables
+    val iOSTarget: (String, KotlinNativeTarget.() -> Unit) -> KotlinNativeTarget =
+        if (System.getenv("SDK_NAME")?.startsWith("iphoneos") == true)
+            ::iosArm64
+        else
+            ::iosX64
+
+    iOSTarget("ios")
+}
+```
+
+```bash
+# workaround 2: make symbolic links to use one source set for two targets
+ln -s iosMain iosArm64Main && ln -s iosMain iosX64Main
+```
+
+Instead of doing this, you can create a hierarchical structure
+with [target shortcuts](multiplatform-share-on-platforms.md#use-target-shortcuts)
+available for typical multi-target scenarios, or you can manually declare and connect the source sets. For example, you
+can create two iOS targets and a shared source set with the `ios()` shortcut:
+
+```kotlin
+kotlin {
+    ios() // iOS device and simulator targets; iosMain and iosTest source sets
+}
+```
+
+The Kotlin toolchain will provide the correct default dependencies and find the API surface area available in the shared
+code. This prevents such cases as, for example, the use of a macOS-specific function in the code shared for Windows.
+
+## For library authors
+
+A hierarchical project structure allows reusing code in similar targets, as well as publishing and consuming libraries
+with granular APIs targeting similar platforms.
+
+The Kotlin toolchain will automatically figure out the API available in the consumer source set while checking the
+unsafe usages, like using an API meant for the JVM in JS code.
+
+* Libraries published with the hierarchical project structure are compatible only with projects that have hierarchical
+  project structure. To enable compatibility with non-hierarchical projects, add the following to
+  the `gradle.properties` file in your library project:
+
+   ```properties
+   kotlin.mpp.enableCompatibilityMetadataVariant=true
```

Comment:
General remark about `enableCompatibilityMetadataVariant`, it is up to you how to deal with it.

Only source code from `commonMain` source set  will be compiled with legacy metadata compiler.

For example in project with such structure: `commonMain -> nativeMain -> [ linuxX64, linuxArm64 ]`
If `enableCompatibilityMetadataVariant` is enabled, then only code in `commonMain` will be compiled with legacy metadata variant.
And both `commonMain` and `nativeMain` will be compiled into klib. 
And the situation goes tricky if platform-specific code will be used in `commonMain`. For example function `getpid` from posix. In such cases, the compilation of `commonMain` source set to legacy format will fail. But if posix's `getpid` will be used in `nativeMain`, everything would compile fine :) 

So that is just an aspect of `enableCompatibilityMetadataVariant` behavior. It is up to you how to work with it.

Reply:
I decided to add a warning on using platform-specific code in `commonMain` in this case

Reply:
I decided to add a warning on using platform-specific code in `commonMain` in this case